### ISSUE TITLE
mesh parameters are not set when using a supplicant

### ISSUE THREAD

**Issue by protectivedad** [openwrt/openwrt#9370]
Branch: v21.02 (probably master also)
Devices: All I think

/lib/netifd/wireless/mac80211.sh (located in packag/kernel/mac80211/files) sets mesh_param parameters from the information in /etc/config/wireless. The variables MP_CONFIG_INT, MP_CONFIG_BOOL and MP_CONFIG_STRING contain the variables that will be set. The problem is that if the mac80211_setup_supplicant is used the script tries to set the variables before the interface is ready. So the iw commands fail which isn't noticed. A solution to this is to wait for the supplicant to have at least setup the interface with a channel and then run the iw command.
The orignal code is:
```
			if [ "$wpa" -gt 0 -o "$auto_channel" -gt 0 ] || chan_is_dfs "$phy" "$channel"; then
				mac80211_setup_supplicant $vif_enable || failed=1
			else
				mac80211_setup_mesh $vif_enable
			fi
		        for var in $MP_CONFIG_INT $MP_CONFIG_BOOL $MP_CONFIG_STRING; do
			        json_get_var mp_val "$var"
			        [ -n "$mp_val" ] && iw dev "$ifname" set mesh_param "$var" "$mp_val"
		        done
```
The updated code is:
```
			if [ "$wpa" -gt 0 -o "$auto_channel" -gt 0 ] || chan_is_dfs "$phy" "$channel"; then
				mac80211_setup_supplicant $vif_enable || failed=1
				[ -z "$failed" ] && \
					while ! iw dev "$ifname" info | grep -q "channel" ; do sleep 1; done
			else
				mac80211_setup_mesh $vif_enable
			fi
			[ -z "$failed" ] && \
				for var in $MP_CONFIG_INT $MP_CONFIG_BOOL $MP_CONFIG_STRING; do
					json_get_var mp_val "$var"
					[ -n "$mp_val" ] && iw dev "$ifname" set mesh_param "$var" "$mp_val"
				done
```
A few thoughts on the solution:
1) If the supplicant never sets up properly this never ends;
Not a problem since without it setting up properly it was never going to be a functional interface anyway. So this doesn't interfere and since the interfaces are run in parallel it doesn't stop the system from coming up and being accessible from ethernet ports.
2) This slows down the startup;
Not really since it is just waiting for a funcational interface anyway. The system is run in parallel so it doesn't slow "access" to the system. When the supplicant has succeeded it will continue on.
3) I assume that if the interface fails there is no point in waiting or setting the variables;
4) I haven't tested all the variables that are set there still maybe some that need to wait longer;
5) There are a few variables that are handled by hostapd/wpa_supplicant I've assumed setting them again is okay (I did have a patch that seperated the two types so as not to set them again, but it seems like extra complication for no real world benefit).

I'd submit my patch but I don't know how :) and it isn't a complicated change.

**Comment by protectivedad** [comment]
After testing some more it works with all almost all the variables. Only mesh_auto_open_plinks and mesh_hwmp_rootmode can't be set without waiting longer.

**Comment by protectivedad** [comment]
Figured out that the order of the variables is important. for example setting mesh_gate_announcements=1 will change mesh_hwmp_rootmode=4. So for mesh_hwmp_rootmode to work it needs to be set last. So:
```
MP_CONFIG_INT="mesh_retry_timeout mesh_confirm_timeout mesh_holding_timeout mesh_max_peer_links
	       mesh_max_retries mesh_ttl mesh_element_ttl mesh_hwmp_max_preq_retries
	       mesh_path_refresh_time mesh_min_discovery_timeout mesh_hwmp_active_path_timeout
	       mesh_hwmp_preq_min_interval mesh_hwmp_net_diameter_traversal_time mesh_hwmp_rootmode
	       mesh_hwmp_rann_interval mesh_gate_announcements mesh_sync_offset_max_neighor
	       mesh_rssi_threshold mesh_hwmp_active_path_to_root_timeout mesh_hwmp_root_interval
	       mesh_hwmp_confirmation_interval mesh_awake_window mesh_plink_timeout"
```
becomes:
```
MP_CONFIG_INT="mesh_retry_timeout mesh_confirm_timeout mesh_holding_timeout mesh_max_peer_links
	       mesh_max_retries mesh_ttl mesh_element_ttl mesh_hwmp_max_preq_retries
	       mesh_path_refresh_time mesh_min_discovery_timeout mesh_hwmp_active_path_timeout
	       mesh_hwmp_preq_min_interval mesh_hwmp_net_diameter_traversal_time
	       mesh_hwmp_rann_interval mesh_gate_announcements mesh_sync_offset_max_neighor
	       mesh_rssi_threshold mesh_hwmp_active_path_to_root_timeout mesh_hwmp_root_interval
	       mesh_hwmp_confirmation_interval mesh_awake_window mesh_plink_timeout mesh_hwmp_rootmode"
```
and I rearranged the order in the for statement.

**Comment by rossiniscarface** [comment]
this also happens in master

**Comment by mundschenk-at** [comment]
> and I rearranged the order in the for statement.

You mean to this?

```bash
for var in $MP_CONFIG_BOOL $MP_CONFIG_STRING $MP_CONFIG_INT; do
```

**Comment by protectivedad** [comment]
> > and I rearranged the order in the for statement.
> 
> You mean to this?
> 
> ```shell
> for var in $MP_CONFIG_BOOL $MP_CONFIG_STRING $MP_CONFIG_INT; do
> ```

No I mean the MP_CONFIG_INT definition above has a var put at the end which eventually expands in the for loop.

**Comment by mundschenk-at** [comment]
Then I misunderstood the "and", I thought you meant separately from the `$MP_CONFIG_INT` line.

**Comment by rossiniscarface** [comment]
@protectivedad can you make a pull request on this? or do you allow that others may?

**Comment by nbd168** [comment]
Fixed in latest master and 23.05
